### PR TITLE
Persist trie nodes to database

### PR DESCRIPTION
**Pull request by ablax**
- [x] Persist trie nodes to database
- [ ] Tests

### CODE REVIEW

File: src/main/java/com/limechain/client/FullNode.java [hunk 1]
```diff
@@ -133,6 +154,7 @@ public List<InsertTrieNode> buildTrieNodesIterator(
         return trieNodesIterator;
     }
 
+
```
Comment (georg-getz) [inline src/main/java/com/limechain/client/FullNode.java]: you can remove one of these lines

File: src/main/java/com/limechain/trie/structure/node/TrieNodeData.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+@EqualsAndHashCode
+public class TrieNodeData {
+    /** The value stored in the trie node. */
+    byte[] value;
+
+    /** A reference to the root of the trie. */
+    byte[] trieRootRef;
+
+    /** The version number of the trie entries. */
+    byte entriesVersion;
```
Comment (georg-getz) [inline src/main/java/com/limechain/trie/structure/node/TrieNodeData.java]: you can remove the comments above the fields as they can be inferred by the field names

File: src/main/java/com/limechain/client/FullNode.java [hunk 3]
```diff
@@ -103,14 +110,28 @@ public List<InsertTrieNode> buildTrieNodesIterator(
...
             byte[] merkleValue = userData.getValue1().get();
+            final InsertTrieNode.InsertStorageValue storageValue = new InsertTrieNode.InsertStorageValue(
+                    storageValueOpt.orElse(null),
+                    storageValueOpt.isPresent(),
+                    false);
 
             byte[] merkleValueCopy = merkleValue.clone();
             List<byte[]> childrenMerkleValues = new ArrayList<>();
-            List<Nibble> partialKeyNibbles = new ArrayList<>(nodeHandle.getPartialKey().asUnmodifiableList());
+            Nibbles partialKey = nodeHandle.getPartialKey();
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/client/FullNode.java:123]: You can use the `Bytes.toArray()` function from `com.google.common.primitives`, that's included, to make it more readable.
`byte[] partialKeyNibbles = Bytes.toArray(partialKey.asUnmodifiableList().stream().map(Nibble::asByte).toList());`

File: src/main/java/com/limechain/trie/structure/node/NodeChildData.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Represents the data associated with a child node in a trie structure.
+ * This class includes the child node's number and its hash.
+ */
+@AllArgsConstructor
+@EqualsAndHashCode
+public class NodeChildData {
+    /** The number assigned to the child node, representing its order. */
+    int childNum;
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/trie/structure/node/NodeChildData.java]: Should these be private?

File: src/main/java/com/limechain/trie/structure/node/TrieNodeData.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Represents the data associated with a node in a trie structure.
+ * This class includes the node's value, a reference to the trie root,
+ * and the version of the entries.
+ */
+@AllArgsConstructor
+@EqualsAndHashCode
+public class TrieNodeData {
+    /** The value stored in the trie node. */
+    byte[] value;
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/trie/structure/node/TrieNodeData.java]: Should these be private?

File: src/main/java/com/limechain/client/FullNode.java [hunk 6]
```diff
@@ -242,4 +264,54 @@ TrieStructure<Pair<Optional<byte[]>, Optional<byte[]>>> initializeDatabaseStorag
...
+     * @param insertTrieNodes The list of trie nodes to be inserted.
+     * @param entriesVersion The version number of the trie entries.
+     */
+    private void insertStorage(KVRepository<String, Object> db, List<InsertTrieNode> insertTrieNodes, byte entriesVersion) {
+        try {
+            for (InsertTrieNode trieNode : insertTrieNodes) {
+                //Insert tn(trie node) to db
+                db.save("tn:" + new String(trieNode.getMerkleValue()), trieNode.getPartialKeyNibbles());
+
+                // Handle storage value
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/client/FullNode.java:270]: Some of the comments in this block can be removed and the code extracted in private methods.

File: src/main/java/com/limechain/client/FullNode.java [hunk 7]
```diff
@@ -242,4 +264,54 @@ TrieStructure<Pair<Optional<byte[]>, Optional<byte[]>>> initializeDatabaseStorag
...
+    private void insertChildren(KVRepository<String, Object> db, InsertTrieNode trieNode)  {
+        for (int childNum = 0; childNum < trieNode.getChildrenMerkleValues().size(); childNum++) {
+            byte[] child = trieNode.getChildrenMerkleValues().get(childNum);
+            NodeChildData nodeChildData = new NodeChildData(childNum, child);
+            //Insert tnc(trie node child) to db
+            db.save("tnc:" + new String(trieNode.getMerkleValue()), nodeChildData);
+        }
+    }
+
+
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/client/FullNode.java]: Remove the multiple empty lines.

**Review by stoyan-lime — APPROVED** [review]
Looks good, added some suggestions.

**Review by stoyan-lime — APPROVED** [review]

**Review by georg-getz — APPROVED** [review]

**Pull request merged**